src/profiling/profiler.cpp

```diff
@@ -1,442 +1,401 @@
-#include <profiling/profiler.hpp>
-#include <util/make_unique.hpp>
-#include <util/debug.hpp>
+#include "profiler.hpp"
 
 namespace arb {
 namespace util {
 
-// Here we provide functionality that the profiler can use to control the CUDA
-// profiler nvprof. The start_nvprof and stop_nvprof calls are provided to let
-// a program control which parts of the program are to be profiled. It is a
-// simple wrapper around the API calls with a mutex to ensure correct behaviour
-// when multiple threads attempt to start or stop the profiler.
-#ifdef ARB_HAVE_GPU
-namespace gpu {
-    bool is_running_nvprof = false;
-    std::mutex gpu_profiler_mutex;
-
-    void start_nvprof() {
-        std::lock_guard<std::mutex> guard(gpu_profiler_mutex);
-        if (!is_running_nvprof) {
-            cudaProfilerStart();
-        }
-        is_running_nvprof = true;
+#ifdef ARB_HAVE_PROFILING
+namespace {
+    // Check whether a string describes a valid profiler region name.
+    bool is_valid_region_string(const std::string& s) {
+        if (s.size()==0u || s.front()=='_' || s.back()=='_') return false;
+        return s.find("__") == s.npos;
     }
 
-    void stop_nvprof() {
-        std::lock_guard<std::mutex> guard(gpu_profiler_mutex);
-        if (is_running_nvprof) {
-            cudaProfilerStop();
+    //
+    // Return a list of the words in the string, using '_' as the delimiter
+    // string, e.g.:
+    //      "communicator"             -> {"communicator"}
+    //      "communicator_events"      -> {"communicator", "events"}
+    //      "communicator_events_sort" -> {"communicator", "events", "sort"}
+    std::vector<std::string> split(const std::string& str) {
+        std::vector<std::string> cont;
+        std::size_t first = 0;
+        std::size_t last = str.find('_');
+        while (last != std::string::npos) {
+            cont.push_back(str.substr(first, last - first));
+            first = last + 1;
+            last = str.find('_', first);
         }
-        is_running_nvprof = false;
+        cont.push_back(str.substr(first, last - first));
+        return cont;
     }
 }
-#else
-namespace gpu {
-    void start_nvprof() {}
-    void stop_nvprof()  {}
-}
-#endif
 
-//
-// profiler_node implementation
-//
-void profiler_node::print(int indent) {
-    std::string s = std::string(indent, ' ') + name;
-    std::cout << s
-              << std::string(60-s.size(), '.')
-              << value
-              << "\n";
-    for (auto& n : children) {
-        n.print(indent+2);
-    }
-}
+// Holds the accumulated number of calls and time spent in a region.
+struct profile_accumulator {
+    std::size_t count=0;
+    double time=0.;
+};
+
+// Records the accumulated time spent in profiler regions on one thread.
+// There is one recorder for each thread.
+class recorder {
+    // used to mark that the recorder is not currently timing a region.
+    static constexpr std::size_t npos = std::numeric_limits<std::size_t>::max();
```

Comment:
-1 is a very obtuse way of saying the largest possible unsigned 64 bit integer.

Reply:
We could argue over this all day! For the record, I think that if you want the largest unsigned integer, it is best to ask for it explicitly.
For what it is worth, the standard also feels the need to explain itself: `This constant is defined with a value of -1, which because size_t is an unsigned integral type, it is the largest possible representable value for this type.`